Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs

```diff
@@ -0,0 +1,333 @@
+            _photoSystem = EntitySystem.Get<PhotoSystem>();
+
+            Dirty();
+        }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(ref _film, "film", 10);
+            serializer.DataField(ref _filmMax, "maxfilm", 10);
+            serializer.DataField(ref _radius, "radius", 1);
+        }
+
+        public override ComponentState GetComponentState()
+        {
+            return new PhotoCameraComponentState(CameraOn, Radius, Film, FilmMax);
+        }
+
+        /// <summary>
+        /// Whether the camera has film or not
+        /// </summary>
+        private bool HasFilm => Film > 0;
+
+        public override void HandleNetworkMessage(ComponentMessage message, INetChannel netChannel, ICommonSession session = null)
+        {
+            base.HandleNetworkMessage(message, netChannel, session);
+
+            switch (message)
+            {
+                case TakingPhotoMessage _:
+                    Film--;
+                    PlaySoundCollection("TakePhoto", -5);
+                    break;
+                case TookPhotoMessage photoTaken:
+                    var author = _entityManager.GetEntity(photoTaken.Author);
+                    var photoEnt = _entityManager.SpawnEntity("Photo", author.Transform.GridPosition);
+                    var photo = photoEnt.GetComponent<PhotoComponent>();
+
+                    Logger.InfoS("photo", $"{author.Name} took a photo at {author.Transform.GridPosition}");
+
+                    if (photoTaken.Suicide)
+                    {
+                        photoEnt.Description = Loc.GetString("The last photo {0} ever took before they went missing...", author.Name);
+
+                        //Drop their items
+                        if (author.TryGetComponent(out IHandsComponent hands))
+                        {
+                            foreach (var heldItem in hands.GetAllHeldItems())
+                            {
+                                hands.Drop(heldItem.Owner, false);
+                            }
+                        }
+
+                        //Drop their gear
+                        if (author.TryGetComponent(out InventoryComponent inv))
+                        {
+                            inv.UnequipAll(true);
+                        }
+
+                        //Vanish without a trace
+                        author.Delete();
+                    } else
+                    {
+                        if(author.TryGetComponent(out HandsComponent hands)){
+                            hands.PutInHand(photoEnt.GetComponent<ItemComponent>());
+                        }
+                    }
+
+                    //Store photo to disk, assign path to photo entity
+                    _photoSystem.StorePhoto(photoTaken.Data, photo);
+
+                    break;
+            }
+        }
+
```

Comment:
Many security holes here:

1. It trusts the client on the author of the photo (can make photos appear in anybody's hands)
2. Doesn't verify that the client is alive/has the camera/camera is on/camera has film.
3. Film and the sound are separate from the actual action of taking the photo meaning that a hacked client could trivially circumvent the film requirement.

A less trivial issue is that the image sent by the client could very easily be malicious. An example is that the size could be something like 100000x100000 of whitespace which could easily crash clients when they try to load it, but compress to "reasonable" sizes. You'll probably have to make the server load the PNG in some sort of safe way to reject it if malicious (hopefully ImageSharp allows inspecting the size without loading it?).

Reply:
I'll Just spawn the photo where the camera is, and if the camera is being held, put it in that entity's hands.
(Ditto for suicides (also also, should I just cut that? it's mildly funny but it shits up the code a bit))

Put some messages in disc about the image bombs, had the same countermeasure thoughts.